[SPARK-14576] [Web UI] Spark console should display Web UI url

seyfe opened the pull request:
## What changes were proposed in this pull request?

This is a proposal to print the Spark Driver UI link when spark-shell is launched.
## How was this patch tested?

Launched spark-shell in local mode and cluster mode. Spark-shell console output included following line:
"Spark context Web UI available at <Spark web url>"

seyfe committed:
Spark console will display UI link once spark context is initialized

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -280,6 +280,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   private[spark] def ui: Option[SparkUI] = _ui
 
+  def uiWebUrl: Option[String] = _ui.map(_.webUrl)
```
I think that's reasonable. I suppose this could be a java.net.URI but no big deal

srowen replied:
PS on second thought maybe we can make this new method private[spark]. Or, how about not adding it? it really doesn't add much in return for adding a whole new API method.

seyfe replied:
`def uiWebUrl: Option[String] ` method is called inside repl's `processLine` so having `private[spark]` won't be enough. It returns a "cannot be accessed" error.

Alternatively, I can change the `def ui: Option[SparkUI]` method from `private[spark]` to `public` and then change the print statement to 
`_sc.ui.foreach(ui => println(s"Spark context Web UI available at ${ui.webUrl}"))`

That would work as well but I preferred to have a new public method (`uiWebUrl `) rather than making the `ui` public. I would be happy to hear your suggestion.

srowen replied:
I understand now why it's done this way. Yes I forgot that this is of course executed in the shell which doesn't have access to private[spark] stuff. Hm, I like the change. I guess I'm a little wary of tacking on yet another method in SparkContext just for it. I wonder if anyone else has any particular feeling o this? a pointer to the UI does sound handy.

srowen commented on core/src/main/scala/org/apache/spark/ui/WebUI.scala:
```diff
@@ -135,17 +135,19 @@ private[spark] abstract class WebUI(
       serverInfo = Some(startJettyServer(host, port, sslOptions, handlers, conf, name))
-      logInfo("Bound %s to %s, and started at http://%s:%d".format(className, host,
-        publicHostName, boundPort))
+      logInfo("Bound %s to %s, and started at %s".format(className, host, webUrl))
```
Here and below you could go ahead and use string interpolation

seyfe replied:
Sure thing. I changed the rest of the file.

srowen commented on repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoopInit.scala:
```diff
@@ -125,6 +125,9 @@ private[repl] trait SparkILoopInit {
       command("""
         @transient val sc = {
           val _sc = org.apache.spark.repl.Main.interp.createSparkContext()
+          _sc.uiWebUrl.map { webUrl =>
```
Nit: `foreach(webUrl => ...)`

seyfe replied:
I addressed this one as well.

seyfe committed:
Addressed the PR feedback:
- Used String interpolation rather than String format. Changed the rest of the file as well.
- Replaced .map with .foreach

srowen commented on core/src/main/scala/org/apache/spark/ui/WebUI.scala:
```diff
@@ -133,26 +133,28 @@ private[spark] abstract class WebUI(
   }
 
+  /** Return the url of web interface. Only valid after bind(). */
+  def webUrl: String = s"http://$publicHostName:$boundPort"
```
Same, doesn't seem like this is used anywhere except a few lines above, so probably doesn't need to be a method?

seyfe replied:
It's used in SparkContext.scala by this method (the one I am adding with this PR)
`def uiWebUrl: Option[String] = _ui.map(_.webUrl)`

andrewor14 commented:
@seyfe please attach a screenshot

srowen commented:
It's pretty much described in the JIRA; it's a tiny change to the shell output. This isn't a change to the web UI actually, but to the link in the shell output that links to the web UI.

srowen commented:
I like it. I wouldn't mind another nod from someone like @andrewor14 if he has a moment to look

seyfe commented:
Hi @andrewor14 . I would like to hear if you have any comments/suggestions or concerns?

srowen commented:
Jenkins test this please

SparkQA commented:
**[Test build #56892 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56892/consoleFull)** for PR 12341 at commit [`f3e8d34`](https://github.com/apache/spark/commit/f3e8d34280dd8c5f212c48920bd0a6460f282cbf).

SparkQA commented:
**[Test build #56892 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56892/consoleFull)** for PR 12341 at commit [`f3e8d34`](https://github.com/apache/spark/commit/f3e8d34280dd8c5f212c48920bd0a6460f282cbf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56892/
Test PASSed.

srowen commented:
@seyfe looks pretty good to me but now needs a rebase, sorry

seyfe commented:
Thanks @srowen. I rebased my change.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #57104 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57104/consoleFull)** for PR 12341 at commit [`ac02b51`](https://github.com/apache/spark/commit/ac02b5176847538a5f45f757a91b7c94f86c16d8).

SparkQA commented:
**[Test build #57104 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57104/consoleFull)** for PR 12341 at commit [`ac02b51`](https://github.com/apache/spark/commit/ac02b5176847538a5f45f757a91b7c94f86c16d8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57104/
Test FAILed.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #57136 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57136/consoleFull)** for PR 12341 at commit [`ac02b51`](https://github.com/apache/spark/commit/ac02b5176847538a5f45f757a91b7c94f86c16d8).

SparkQA commented:
**[Test build #57136 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57136/consoleFull)** for PR 12341 at commit [`ac02b51`](https://github.com/apache/spark/commit/ac02b5176847538a5f45f757a91b7c94f86c16d8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57136/
Test FAILed.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #57240 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57240/consoleFull)** for PR 12341 at commit [`ac02b51`](https://github.com/apache/spark/commit/ac02b5176847538a5f45f757a91b7c94f86c16d8).

SparkQA commented:
**[Test build #57240 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57240/consoleFull)** for PR 12341 at commit [`ac02b51`](https://github.com/apache/spark/commit/ac02b5176847538a5f45f757a91b7c94f86c16d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57240/
Test PASSed.

srowen commented:
Merged to master
